[BUG] Writing Parquet map(map) column can not set the outer key as non-null.

**Describe the bug**
Reports the following error when writing map(map) column.
```
key column can not be nullable
```

The SchemaUtils.writerOptionsFromField can not handle map(map) column.
`SchemaUtils` uses `structBuilder` to simulate mapBuilder, but does not set the outer key as non-null. For the inner key, it's OK.

**Steps/Code to reproduce bug**
```
val data = Seq(
  (Map(Map(111->111, 112->112) -> 1, Map(121->121, 122->122) -> 2), 1),
  (Map(Map(211->111, 212->112) -> 1, Map(221->121, 222->122) -> 2), 2)
)
val df = spark.createDataFrame(data).toDF("c1", "c2")
df.write.parquet("/tmp/a.parquet")
```

Error:
```
Caused by: org.apache.spark.SparkException: Job aborted due to stage failure: Task 0 in stage 0.0 failed 1 times, most recent failure: Lost task 0.0 in stage 0.0 (TID 0) (chongg-pc executor driver): java.lang.IllegalArgumentException: key column can not be nullable
	at ai.rapids.cudf.ColumnWriterOptions.mapColumn(ColumnWriterOptions.java:530)
	at com.nvidia.spark.rapids.SchemaUtils$.writerOptionsFromField(SchemaUtils.scala:298)
	at com.nvidia.spark.rapids.SchemaUtils$.$anonfun$writerOptionsFromSchema$1(SchemaUtils.scala:329)
	at scala.collection.Iterator.foreach(Iterator.scala:943)
	at scala.collection.Iterator.foreach$(Iterator.scala:943)
	at scala.collection.AbstractIterator.foreach(Iterator.scala:1431)
	at scala.collection.IterableLike.foreach(IterableLike.scala:74)
	at scala.collection.IterableLike.foreach$(IterableLike.scala:73)
	at org.apache.spark.sql.types.StructType.foreach(StructType.scala:102)
	at com.nvidia.spark.rapids.SchemaUtils$.writerOptionsFromSchema(SchemaUtils.scala:327)
	at com.nvidia.spark.rapids.GpuParquetWriter.<init>(GpuParquetFileFormat.scala:374)
	at com.nvidia.spark.rapids.GpuParquetFileFormat$$anon$1.newInstance(GpuParquetFileFormat.scala:287)
	at org.apache.spark.sql.rapids.GpuSingleDirectoryDataWriter.newOutputWriter(GpuFileFormatDataWriter.scala:235)
	at org.apache.spark.sql.rapids.GpuSingleDirectoryDataWriter.<init>(GpuFileFormatDataWriter.scala:217)
	at org.apache.spark.sql.rapids.GpuFileFormatWriter$.executeTask(GpuFileFormatWriter.scala:326)
	at org.apache.spark.sql.rapids.GpuFileFormatWriter$.$anonfun$write$15(GpuFileFormatWriter.scala:266)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:90)
	at org.apache.spark.scheduler.Task.run(Task.scala:136)
	at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$3(Executor.scala:548)
	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1504)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:551)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:750)

```

**Environment details (please complete the following information)**
Branch 23.10

--- a/sql-plugin/src/main/scala/com/nvidia/spark/rapids/SchemaUtils.scala
+++ b/sql-plugin/src/main/scala/com/nvidia/spark/rapids/SchemaUtils.scala
@@ -35,7 +35,7 @@ import org.apache.spark.sql.types._
 
 object SchemaUtils {
   // Parquet field ID metadata key
-  val FIELD_ID_METADATA_KEY = "parquet.field.id"
+  private val FIELD_ID_METADATA_KEY = "parquet.field.id"
 
   /**
    * Convert a TypeDescription to a Catalyst StructType.
@@ -236,7 +236,8 @@ object SchemaUtils {
       nullable: Boolean,
       writeInt96: Boolean,
       fieldMeta: Metadata,
-      parquetFieldIdWriteEnabled: Boolean): T = {
+      parquetFieldIdWriteEnabled: Boolean,
+      isMapKey: Boolean): T = {
 
     // Parquet specific field id
     val parquetFieldId: Option[Int] = if (fieldMeta.contains(FIELD_ID_METADATA_KEY)) {
@@ -275,19 +276,25 @@ object SchemaUtils {
             a.elementType,
             name,
             a.containsNull,
-            writeInt96, fieldMeta, parquetFieldIdWriteEnabled).build())
+            writeInt96, fieldMeta, parquetFieldIdWriteEnabled, isMapKey = false).build())
       case m: MapType =>
+        // if this is a key of another map. e.g.: map1(map2(int,int), int), the map2 is the map
+        // key of map1, map2 should be non-nullable
+        val isMapNullable = if (isMapKey) false else nullable
+
         // It is ok to use `StructBuilder` here for key and value, since either
         // `OrcWriterOptions.Builder` or `ParquetWriterOptions.Builder` is actually an
         // `AbstractStructBuilder`, and here only handles the common column metadata things.
         builder.withMapColumn(
           mapColumn(name,
             writerOptionsFromField(
+              // This nullable is useless because we use the child of struct column
               structBuilder(name, nullable),
               m.keyType,
               "key",
               nullable = false,
-              writeInt96, fieldMeta, parquetFieldIdWriteEnabled).build().getChildColumnOptions()(0),
+              writeInt96, fieldMeta, parquetFieldIdWriteEnabled, isMapKey = true).build()
+                .getChildColumnOptions()(0),
             writerOptionsFromField(
               structBuilder(name, nullable),
               m.valueType,
@@ -295,7 +302,9 @@ object SchemaUtils {
               m.valueContainsNull,
               writeInt96,
               fieldMeta,
-              parquetFieldIdWriteEnabled).build().getChildColumnOptions()(0)))
+              parquetFieldIdWriteEnabled, isMapKey = false).build().getChildColumnOptions()(0),
+            // set the nullable for this map
+            isMapNullable))
       case BinaryType =>
         if (parquetFieldIdWriteEnabled && parquetFieldId.nonEmpty) {
           builder.withBinaryColumn(name, nullable, parquetFieldId.get)
@@ -326,7 +335,7 @@ object SchemaUtils {
       parquetFieldIdEnabled: Boolean = false): T = {
     schema.foreach(field =>
       writerOptionsFromField(builder, field.dataType, field.name, field.nullable, writeInt96,
-        field.metadata, parquetFieldIdEnabled)
+        field.metadata, parquetFieldIdEnabled, isMapKey = false)
     )
     builder.asInstanceOf[T]
   }

--- a/sql-plugin/src/main/scala/com/nvidia/spark/rapids/SchemaUtils.scala
+++ b/sql-plugin/src/main/scala/com/nvidia/spark/rapids/SchemaUtils.scala
@@ -236,8 +236,7 @@ object SchemaUtils {
       nullable: Boolean,
       writeInt96: Boolean,
       fieldMeta: Metadata,
-      parquetFieldIdWriteEnabled: Boolean,
-      isMapKey: Boolean): T = {
+      parquetFieldIdWriteEnabled: Boolean): T = {
 
     // Parquet specific field id
     val parquetFieldId: Option[Int] = if (fieldMeta.contains(FIELD_ID_METADATA_KEY)) {
@@ -276,12 +275,8 @@ object SchemaUtils {
             a.elementType,
             name,
             a.containsNull,
-            writeInt96, fieldMeta, parquetFieldIdWriteEnabled, isMapKey = false).build())
+            writeInt96, fieldMeta, parquetFieldIdWriteEnabled).build())
       case m: MapType =>
-        // if this is a key of another map. e.g.: map1(map2(int,int), int), the map2 is the map
-        // key of map1, map2 should be non-nullable
-        val isMapNullable = if (isMapKey) false else nullable
-
         // It is ok to use `StructBuilder` here for key and value, since either
         // `OrcWriterOptions.Builder` or `ParquetWriterOptions.Builder` is actually an
         // `AbstractStructBuilder`, and here only handles the common column metadata things.
@@ -293,8 +288,7 @@ object SchemaUtils {
               m.keyType,
               "key",
               nullable = false,
-              writeInt96, fieldMeta, parquetFieldIdWriteEnabled, isMapKey = true).build()
-                .getChildColumnOptions()(0),
+              writeInt96, fieldMeta, parquetFieldIdWriteEnabled).build().getChildColumnOptions()(0),
             writerOptionsFromField(
               structBuilder(name, nullable),
               m.valueType,
@@ -302,9 +296,12 @@ object SchemaUtils {
               m.valueContainsNull,
               writeInt96,
               fieldMeta,
-              parquetFieldIdWriteEnabled, isMapKey = false).build().getChildColumnOptions()(0),
+              parquetFieldIdWriteEnabled).build().getChildColumnOptions()(0),
             // set the nullable for this map
-            isMapNullable))
+            // if `m` is a key of another map, this `nullable` should be false
+            // e.g.: map1(map2(int,int), int), the map2 is the map
+            // key of map1, map2 should be non-nullable
+            nullable))
       case BinaryType =>
         if (parquetFieldIdWriteEnabled && parquetFieldId.nonEmpty) {
           builder.withBinaryColumn(name, nullable, parquetFieldId.get)
@@ -335,7 +332,7 @@ object SchemaUtils {
       parquetFieldIdEnabled: Boolean = false): T = {
     schema.foreach(field =>
       writerOptionsFromField(builder, field.dataType, field.name, field.nullable, writeInt96,
-        field.metadata, parquetFieldIdEnabled, isMapKey = false)
+        field.metadata, parquetFieldIdEnabled)
     )
     builder.asInstanceOf[T]
   }

